### ISSUE TITLE
Inference for gene expression of Fetal Astrocyte 1

### ISSUE THREAD

**Issue by tristan-ka** [GET-Foundation/get_model#21]
Hi again,  

I am trying to reproduce the results presented in the [Hugging Face space](https://huggingface.co/spaces/get-foundation/getdemo). My goal is to call the model in inference on a specific region and extract the gene expression prediction.  

I attempted this using the **Fetal Astrocyte 1** cell type and wrote the following script to perform the prediction. However, I observed **significant discrepancies** between my predictions and the ground truth.  

## Questions  

1. **Model Selection**  
   - In the provided S3 bucket, which model checkpoint should I use? I am currently using:  
     ```
     s3://2023-get-xf2217/get_demo/checkpoints/fetal_pretrain_fetal_finetune/Astrocytes/checkpoint-best.pth
     ```
     Is this the correct one for my use case?  
    - When loading the model, I need to perform a significant amount of layer renaming. Is that normal? 
2. **Post-Processing for Ground Truth**  
   - To align my results with the expected ground truth, I applied `log1p` transformation to the gene expression values.  
   - Are there any additional post-processing steps required?  
   - To better understand the ground truth gene expression, should I look at `_make_dataset` in `zarr_dataset.py`?  

3. **Reproducibility**  
   - Would you be able to provide a **detailed notebook** demonstrating a simple gene expression prediction workflow for a given cell type?  

## Implementation  

```Python
import hydra
import os
import torch
import pandas as pd
import numpy as np

from hydra.core.global_hydra import GlobalHydra
from hydra.utils import instantiate

# path with all genomics data -> s3://2023-get-xf2217/get_demo/pretrain_human_bingren_shendure_apr2023/fetal_adult/
data_path = 'paths_with_all_genomics_data' 

# the checkpoint is from astrocytes cells
checkpoint_path = os.path.join(data_path, 'astrocyte-checkpoint-best.pth') # downloaded from s3://2023-get-xf2217/get_demo/checkpoints/fetal_pretrain_fetal_finetune/Astrocytes/checkpoint-best.pth

checkpoint_dict = torch.load(checkpoint_path, map_location='cpu')
pytorch_params = checkpoint_dict['model']

GlobalHydra.instance().clear()
hydra.initialize(config_path="../get_model/get_model/config/model")
cfg = hydra.compose(config_name="GETRegionFinetune.yaml")

model = instantiate(cfg.model)

mapping = {
    "cls_token": "cls_token",
    "region_embed.embed.weight": "region_embed.proj.weight",
    "region_embed.embed.bias": "region_embed.proj.bias",
    "encoder.norm.weight": "fc_norm.weight",
    "encoder.norm.bias": "fc_norm.bias",
    "head_exp.head.weight": "head.weight",
    "head_exp.head.bias": "head.bias"
}

# Dynamically add blocks mappings
for i in range(12):
    for param in ["norm1.weight", "norm1.bias", "attn.q_bias", "attn.v_bias", "attn.qkv.weight", "attn.proj.weight", "attn.proj.bias", "norm2.weight", "norm2.bias", "mlp.fc1.weight", "mlp.fc1.bias", "mlp.fc2.weight", "mlp.fc2.bias"]:
        mapping[f"encoder.blocks.{i}.{param}"] = f"blocks.{i}.{param}"

for new_key, old_key in mapping.items():
    if old_key in pytorch_params:
        pytorch_params[new_key] = pytorch_params.pop(old_key)
    else:
        raise ValueError(f"Key {old_key} not found in checkpoint")
    
model.load_state_dict(pytorch_params)


cell_df = pd.read_csv(os.path.join(data_path, 'cell_type_pretrain_human_bingren_shendure_apr2023.csv'), delimiter=";")
id_cell = cell_df[(cell_df['expression'] == True) & (cell_df['cluster'] == 'astrocyte')]['id'].values[0]

gene_exp_df = pd.read_feather(os.path.join(data_path, f'genomics_data/{id_cell}.exp.feather' ))
motifs_df = pd.read_feather(os.path.join(data_path, f'genomics_data/{id_cell}.atac.motif.output.feather' ))

gene_exp_df['log1p_tpm'] = np.log1p(gene_exp_df['TPM'])

#select a region to predict gene expression
region_index = 156

ground_truth_exp = gene_exp_df['log1p_tpm'][(gene_exp_df['index']==region_index) & (gene_exp_df['Strand']=="+")].values # only looking at positive strand for now
print('Ground truth as average:', ground_truth_exp.mean())
print('ground truth as max', ground_truth_exp.max())
input_motifs = motifs_df.iloc[region_index].values[4:]
input_motifs = np.array(input_motifs, dtype=np.float32)
input_motifs = torch.tensor(input_motifs, dtype=torch.float32).reshape(1,1,283)
model.eval()
predicted_gene_exp = model(input_motifs)[0][0]
print('predicted gene expression for single region input:', predicted_gene_exp.detach().numpy()[0])

if region_index > 100:
    indices = range(region_index-100,region_index+100)
    out_index = 100
else:
    indices = range(0,200)
    out_index = region_index
    

input_motifs = motifs_df.iloc[indices].values[:,4:]

input_motifs = np.array(input_motifs, dtype=np.float32)
input_motifs = torch.tensor(input_motifs, dtype=torch.float32).reshape(1,200,283)
model.eval()
predicted_gene_exp = model(input_motifs)[0][out_index]
print('Predicted gene expression for 200 regions input:', predicted_gene_exp.detach().numpy()[0])
```

**Comment by fuxialexander** [comment]
Hi, the one you should use is `aws s3 ls s3://2023-get-xf2217/get_demo/checkpoints/regulatory_inference_checkpoint_fetal_adult/finetune_fetal_adult_leaveout_astrocyte/checkpoint-best.pth` and set `cfg.dataset.quantitative_atac` to `false`.

The one you currently use is for using with an older version of data which was based on peaks called in the fetal atlas paper and use 900 region per sample.

**Comment by fuxialexander** [comment]
This would probably work with the finetune_tutorial config when using the predict stage: https://zenodo.org/records/14614947/files/astrocyte.zarr.tar?download=1

**Comment by fuxialexander** [comment]
I just updated a tutorial Jupyter notebook to do that. Please checkout https://github.com/GET-Foundation/get_model/blob/master/tutorials/infer_astrocyte.ipynb

**Comment by tristan-ka** [comment]
Hi,  

Thanks a lot for sharing this notebook! I'm currently working on refining my understanding of the inference process, and your work has been incredibly helpful. It allowed me to extract the core calls outside of the Lightning dependencies.  

I've put together the following piece of code based on that:  

```Python
import torch
import logging
import numpy as np

import matplotlib.pyplot as plt
from scipy.stats import pearsonr
from hydra.utils import instantiate

from get_model.dataset.zarr_dataset import RegionMotifDataset
from get_model.config.config import load_config_from_yaml
from get_model.utils import load_checkpoint, extract_state_dict, rename_state_dict, load_state_dict

def get_model(cfg):
    model = instantiate(cfg.model)

    # Load main model checkpoint
    if cfg.finetune.checkpoint is not None:
        checkpoint_model = load_checkpoint(
            cfg.finetune.checkpoint, model_key=cfg.finetune.model_key
        )
        checkpoint_model = extract_state_dict(checkpoint_model)
        checkpoint_model = rename_state_dict(
            checkpoint_model, cfg.finetune.rename_config
        )
        load_state_dict(
            model, checkpoint_model, strict=cfg.finetune.strict
        )

    # Load additional checkpoints
    if len(cfg.finetune.additional_checkpoints) > 0:
        for checkpoint_config in cfg.finetune.additional_checkpoints:
            checkpoint_model = load_checkpoint(
                checkpoint_config.checkpoint, model_key=checkpoint_config.model_key
            )
            checkpoint_model = extract_state_dict(checkpoint_model)
            checkpoint_model = rename_state_dict(
                checkpoint_model, checkpoint_config.rename_config
            )
            load_state_dict(
                model, checkpoint_model, strict=checkpoint_config.strict
            )

    model.freeze_layers(
        patterns_to_freeze=cfg.finetune.patterns_to_freeze, invert_match=False
    )
    logging.debug("Model = %s" % str(model))
    return model


if __name__ == '__main__':
    # load the config from the yaml file
    cfg = load_config_from_yaml("exported_training_config.yaml")
    cfg.stage='validate'

    # Load dataset
    dataset = RegionMotifDataset(
        **cfg.dataset,
        is_train=False,
    )

    # Create DataLoader
    dataloader = torch.utils.data.DataLoader(
        dataset,
        batch_size=cfg.machine.batch_size,
        num_workers=cfg.machine.num_workers,
        drop_last=False,
        shuffle=False,
    )

    # Load model
    model = get_model(cfg)
    model.eval()

    results = []
    labels = []
    with torch.no_grad():
        for batch in dataloader:
            outputs = model(batch['region_motif']).numpy().tolist()
            results.extend(outputs)
            labels.extend(batch['exp_label'].numpy().tolist())
    results = np.array(results).reshape(-1,2)
    labels = np.array(labels).reshape(-1,2)

    plt.figure()
    plt.scatter(results[:,0], labels[:,0], label='positive strand')
    plt.scatter(results[:,1], labels[:,1], label='negative strand')
    plt.legend()
    plt.xlabel('predicted')
    plt.ylabel('labels')


    print(pearsonr(results[:,0], labels[:,0]))
    print(pearsonr(results[:,1], labels[:,1]))
``` 

Using this approach, I’m able to obtain promising predictions. However, I haven’t been able to match the Pearson correlation you reported—I currently get 0.68.

![Image](https://github.com/user-attachments/assets/1099879a-e12b-4c51-b04f-a95cb246ee6e)

Is there any post-processing applied to the results that might explain the difference?

**Comment by fuxialexander** [comment]
That's weird, which data are you using? Are you able to replicate the result in the infer_astrocyte notebook? Here it looks that some expression label was set to zero for some reason.

**Comment by tristan-ka** [comment]
Yes, I was able to replicate the results using the notebook you provided, which confirms that the data files I’m using are correct.

You’re right - most of the labels in the dataset are zeros. When I exclude them, the Pearson correlation improves significantly (0.92 for positive and 0.95 for negative strand).

```python
zero_indices = np.concatenate([np.where(labels[:,0] == 0)[0], np.where(labels[:,1] == 0)[0]])
labels = np.delete(labels, zero_indices, axis=0)
results = np.delete(results, zero_indices, axis=0)
```

In total, this removes 10,984 out of 11,200 data points, leaving only 216 gene expression predictions.

This suggests that the issue might be related to the data loader. Another possibility is that making calls from outside the Lightning modules is affecting the results. 

### Dataset Loading Code  

Here’s the code I’m using to load the dataset and create the DataLoader. It loads the configuration from a YAML file, initializes the `RegionMotifDataset` with the validation stage, and sets up a DataLoader with the specified batch size and number of workers.  

```python
# load the config from the yaml file
cfg = load_config_from_yaml("exported_training_config.yaml")
cfg.stage='validate'

# Load dataset
dataset = RegionMotifDataset(
    **cfg.dataset,
    is_train=False,
)

# Create DataLoader
dataloader = torch.utils.data.DataLoader(
    dataset,
    batch_size=cfg.machine.batch_size,
    num_workers=cfg.machine.num_workers,
    drop_last=True,
    persistent_workers=True,
)
```

I haven’t found a definitive answer yet and debugging is not straightforward within the encapsulated structure of Lightning code. 

### Strange Behavior with Batch Size  

I've also noticed an unexpected behavior in my code. Changing the batch size significantly alters the number of data points. For instance, reducing the batch size from 8 to 4 reduces the dataset size from 11,200 to 10,400.  

Could this be due to the random shift in the dataset creation within the `setup` function of `RegionMotifDataset`?  

### Questions That Might Help Debug  

- Do you fix the random seed when generating the validation set?  
- What is the exact size of your validation set? 

Thanks again for your reactivity 🙏

**Comment by fuxialexander** [comment]
Sorry I’m traveling until Saturday. Will take a look when I get back.

>In total, this removes 10,984 out of 11,200 data points, leaving only 216 gene expression predictions.

How are these labels constructed?

**Comment by tristan-ka** [comment]
I’m using the `RegionMotifDataset` class in your code with the exact same configuration as in the inference notebook you provided. The cfg.dataset.zarr_path is set to the path of the astrocytes Zarr folder that I downloaded using the notebook.

**Comment by fuxialexander** [comment]
I see what happened. You are plotting all the region's prediction with label while in fact not all of them are promoters. The meaningful labels are those in the promoter and we just set other region to have 0 labels as scRNA-seq won't measure them. In each batch, there is a batch['mask'] which tells you which region idx is a promoter. Try below:

# Load dataset
dataset = RegionMotifDataset(
    **cfg.dataset,
    is_train=False,
)

# Create DataLoader
dataloader = torch.utils.data.DataLoader(
    dataset,
    batch_size=cfg.machine.batch_size,
    num_workers=cfg.machine.num_workers,
    drop_last=False,
    shuffle=False,
)

# Load model
model = get_model(cfg)
model.eval()

results = []
labels = []
with torch.no_grad():
    for batch in dataloader:
        tss_mask = batch['mask'].numpy()
        outputs = model(batch['region_motif']).numpy()[tss_mask>0]
        if len(outputs)>0:
            results.extend(outputs)
            labels.extend(batch['exp_label'].numpy()[tss_mask>0])
results = np.array(results)
labels = np.array(labels)
plt.figure()
plt.scatter(results, labels, label='Both strand')
plt.legend()
plt.xlabel('predicted')
plt.ylabel('labels')

print(pearsonr(results, labels))

**Comment by tristan-ka** [comment]
Thanks a lot for going through my code and providing a solution.
